share/ipld/get.go

```diff
@@ -235,7 +273,7 @@ func GetLeavesByNamespace(
 
```

Comment:
commenting here because Github doesn't let you comment on code outside of the changes...

[non-blocking]
Unless I'm missing something, we are calling `wg.add()`, significantly more that `wg.done()`
We have one `defer wg.done()` call here but we are calling `wg.add()` in a loop here https://github.com/celestiaorg/celestia-node/pull/1339/files#diff-2ff88a4977c65c215324c40905aa73d0f8a9e09c6715437e058dc0fa95d2c478R335

additionally, we start off with a `wg.add` call here https://github.com/celestiaorg/celestia-node/pull/1339/files#diff-2ff88a4977c65c215324c40905aa73d0f8a9e09c6715437e058dc0fa95d2c478R217 with no corresponding `wg.done` call. 

Where are the rest of the `wg.done` calls?

I know this is out of scope of this PR, I'd just like to understand the flow here better.

Reply:
this comment needs more context imo https://github.com/celestiaorg/celestia-node/pull/1339/files#diff-2ff88a4977c65c215324c40905aa73d0f8a9e09c6715437e058dc0fa95d2c478R334

Reply:
Unfortunately, links you've shared are not working for me

Reply:
I got the question, though.

Reply:
All `wg.add`s happen for each CID/hash in the tree. They are added right before they are queued and `wg,done`ed in defer as the end of the  processing.

The root `wg.add` is called for the first root CID before processing starts. While being processed, we extract two child links/cids/hashes, `wg.add` and queue them, and lastly the defer`wg.done` for the root is invoked. After the newle extracted links are processed in the same way, with the respective defers being called. Altogether, the amount of `adds` and `dones` is always the same.

Reply:
@Wondertan explained it well. I want to add small detail that WaitGroup is a counter for amount of goruotine running in parallel. For every node being processed by workerpool a `Job` is created. When node is processed and goroutine exits it calls `Done`.